### DISCUSSION TITLE
Hires fix - scale bigger, not sure if it's worth sharing but...

### DISCUSSION THREAD

**Discussion by Seafan75** [vladmandic/sdnext#795] [General]
I was kind of bored and was pretty sure that the hires fix had better memory handling before.
 
So, without really knowing much python (and almost 0 regarding pytorch and so on), I just did some step through debugging.
And it seems to have worked, somewhat.

Running on a 4090, I could run a batch of 512*1024 * 4 and hires fix it with 1.8 scale. 1.85 scale got me an memory error.
After the change I could run the same batch with 2.05 scale. So 10-15-ish percent better.

I did try back and forth a bunch, but it seems consistant and doesn't seem to break anything. 
Technically this isn't a part of this repo (it is in [stability-AI](https://github.com/Stability-AI/stablediffusion/blob/main/ldm/modules/diffusionmodules/model.py)), but it seems dead and I just wanted to share.

The changes, in the repositories\stable-diffusion-stability-ai\ldm\modules\diffusionmodules\model.py:
After the 2 rows that goes "h = hs[-1]", just add "del hs" to remove it from memory (I guess).
Also adding "torch.cuda.empty_cache()" gave me an additional 0.05 scale, but I think that might be a bit random.

If anyone feels to try it, I would be curious if you get similar results.

**Comment by vladmandic** [comment]
Interesting, I'll take look.

Btw, anytime you ask cuda to do GC, it has to pause to feeze current state and figure out what needs to be done. So I'm doing forced GC only when things are not actually running.

**Comment by elen07zz** [comment]
> I was kind of bored and was pretty sure that the hires fix had better memory handling before.
> 
> So, without really knowing much python (and almost 0 regarding pytorch and so on), I just did some step through debugging. And it seems to have worked, somewhat.
> 
> Running on a 4090, I could run a batch of 512*1024 * 4 and hires fix it with 1.8 scale. 1.85 scale got me an memory error. After the change I could run the same batch with 2.05 scale. So 10-15-ish percent better.
> 
> I did try back and forth a bunch, but it seems consistant and doesn't seem to break anything. Technically this isn't a part of this repo (it is in [stability-AI](https://github.com/Stability-AI/stablediffusion/blob/main/ldm/modules/diffusionmodules/model.py)), but it seems dead and I just wanted to share.
> 
> The changes, in the repositories\stable-diffusion-stability-ai\ldm\modules\diffusionmodules\model.py: After the 2 rows that goes "h = hs[-1]", just add "del hs" to remove it from memory (I guess). Also adding "torch.cuda.empty_cache()" gave me an additional 0.05 scale, but I think that might be a bit random.
> 
> If anyone feels to try it, I would be curious if you get similar results.

which line exactly? i think they are two h=hs[-1]
there is one on line 427 and another on line 532

**Reply by Seafan75** [reply]
Did some more testing, this seems to work "as intended" (for me, with my specific settings).
Replace the function forward on row 518 with:

```
    def forward(self, x):
        # timestep embedding
        temb = None

        # downsampling
        hs = [self.conv_in(x)]
        del x
        for i_level in range(self.num_resolutions):
            for i_block in range(self.num_res_blocks):
                if len(self.down[i_level].attn) > 0:
                    h = self.down[i_level].attn[i_block](h)
                else:
                    h = self.down[i_level].block[i_block](hs[-1], temb)
                hs = [h]
            if i_level != self.num_resolutions-1:
                hs = [(self.down[i_level].downsample(hs[-1]))]

        # middle
        hs[-1] = self.mid.block_1(hs[-1], temb)
        hs[-1] = self.mid.attn_1(hs[-1])
        hs[-1] = self.mid.block_2(hs[-1], temb)

        # end
        hs[-1] = self.norm_out(hs[-1])
        hs[-1] = nonlinearity(hs[-1])
        hs[-1] = self.conv_out(hs[-1])
        return hs[-1]

```
My testing is basically to increase hires fix (text2img) until it breaks. Note down the scale. Kill the server, add the "fix", do the same thing again.

**Comment by vladmandic** [comment]
the issue is that `repositories\stable-diffusion-stability-ai` is an external repo (as is anything in `/repositories`) - thats the original code that a1111 is built on top of, if i modify anything there, i'd have to import it completely instead of linking to it. i'll think about it, perhaps its time to do that.

**Comment by vladmandic** [comment]
i took a second look at this and it may cause some issues with model training as value wont be available for gradient calculation in the backward pass. risk is not worth it.
